src/main/java/org/jabref/logic/exporter/CffExporter.java

```diff
@@ -0,0 +1,242 @@
+package org.jabref.logic.exporter;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.time.LocalDate;
+import java.time.Year;
+import java.time.YearMonth;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import org.jabref.logic.util.OS;
+import org.jabref.logic.util.StandardFileType;
+import org.jabref.model.database.BibDatabaseContext;
+import org.jabref.model.entry.Author;
+import org.jabref.model.entry.AuthorList;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.field.BiblatexSoftwareField;
+import org.jabref.model.entry.field.Field;
+import org.jabref.model.entry.field.StandardField;
+import org.jabref.model.entry.field.UnknownField;
+import org.jabref.model.entry.types.EntryType;
+import org.jabref.model.entry.types.StandardEntryType;
+
+/**
+ * Exporter for exporting in CFF format.
+ */
+class CffExporter extends Exporter {
+    public CffExporter() {
+        super("cff", "CFF", StandardFileType.CFF);
+    }
+
+    @Override
+    public void export(BibDatabaseContext databaseContext, Path file, List<BibEntry> entries) throws Exception {
```

Comment:
Why is a complete new Exporter needed? You know that you can adjust the template based exporter for different entry types as well?

Reply:
This is more complicated than just converting JabRef's entry types to CFF's entry types. Using a template-based exporter, I can only use formatters which take a `String` as an input and render another `String`. This is the way I  first implemented it in #10917, but when I wanted to improve it I faced to many obstacles. 

This exporter needs to know if it needs a `preferred-citation` field (i.e. if the input BibEntry is a type different than `software` or `dataset`), and in this case it must adapt its behavior a lot, and change the display of other fields than the type accordingly. This is complicated to do using a template-based exporter, because I can't pass the whole `BibEntry` as input to the formatter, but only a `String` corresponding to only one field which does not know nothing about the rest. To summarize, the exporter from JabRef's format to CFF format requires too much logic to be handled by a template-based exporter. 

Therefore, I chose to replace the existing template-based exporter by a whole new Exporter. This was a possibility  proposed by @koppor's [comment](https://github.com/JabRef/jabref/issues/10661#issuecomment-1830488166) : 
> The export can either be done by Java code directly or creating .layout files.

Reply:
Somehow refs https://github.com/koppor/jabref/issues/392 (which also is a possible GSoC project)

Reply:
Somehow related: "preferred citation" is a relation of a bibentry to another bibentry. Meaning: One CFF will be imported into 2 bibentries and 2 bibentries will be written to 1 CFF (the software + preferred citation). Refs https://github.com/koppor/jabref/issues/14#issuecomment-1751841346.

Seeing `cites` and `related` being the only option, I say: use `cites` (from software/dataset) A to another entry B. This one interpreted as "preferred-citation"  when exporting. When B is exported as CFF, A is exprted as main block and B as "preferred-ciation".

Reply:
CFF standard: https://github.com/citation-file-format/citation-file-format/blob/main/schema-guide.md

Reply:
We discussed that the template exporter has some magic for auto-choosing the right template according to the type. one could have software and data as top yaml and the others as the nesting. However, since there is preferred-citation and `references`, the distinction is not only with types, but also with the relations. IMHO our [custom export filters](https://docs.jabref.org/collaborative-work/export/customexports) do not have the expressive power. Thus, one really needs to craft own code. Please use a YAML library such as [snakeyaml](https://bitbucket.org/snakeyaml/snakeyaml-engine/src/master). 

![image](https://github.com/JabRef/jabref/assets/1366654/61d08cd0-dbaa-4b61-bcea-904b7eb61a38)

Reply:
> Somehow refs [koppor#392](https://github.com/koppor/jabref/issues/392) (which also is a possible GSoC project)

I would possibly be interested to conduct this GSoC project !

> Somehow related: "preferred citation" is a relation of a bibentry to another bibentry. Meaning: One CFF will be imported into 2 bibentries and 2 bibentries will be written to 1 CFF (the software + preferred citation). Refs [koppor#14 (comment)](https://github.com/koppor/jabref/issues/14#issuecomment-1751841346).
> 
> Seeing `cites` and `related` being the only option, I say: use `cites` (from software/dataset) A to another entry B. This one interpreted as "preferred-citation" when exporting. When B is exported as CFF, A is exprted as main block and B as "preferred-ciation".

I have several questions regarding this. 

+ Firstly, when the user wants to export an entry **B** with a type that is not "software" nor "dataset" to CFF, do you mean I export to a dummy "main" block **A** with a `cites` field actually referencing this entry **B** ? This is what I am actually doing in this PR, but using `preferred-citation` instead of `cites`. The entry **B** does not know which entry **A** is citing it, that's why if only **B** is exported there is no other choice than having a dummy entry **A**.
+ Secondly, how to export **2** bibentries to 1 CFF if both of these entries are from a type different than "software" or "dataset" ?
+ Thirdly, do you mean we have to deduct the link between the first and the second entry using the JabRef's bibentry's field `cites` ? I mean, if there are two entries with one citing the other passed to the `export` function, then we only export one CFF file ? 
+ Fourthly, do you want to modify the importer so that it uses JabRef's filed `cites` to parse the `preferred-citation` field ? This way, if a CFF file to be imported contains a `preferred-citation` field, then the importer will create two entries: one for the software which cites the other one, for the `preferred-citation`.

Reply:
> `references` in CFF are mirrored in JabRef as `related`

Regarding this, you would like to parse the `references` field in both the importer and the exporter ?

Reply:
> We discussed that the template exporter has some magic for auto-choosing the right template according to the type. one could have software and data as top yaml and the others as the nesting. However, since there is preferred-citation and `references`, the distinction is not only with types, but also with the relations. IMHO our [custom export filters](https://docs.jabref.org/collaborative-work/export/customexports) do not have the expressive power. Thus, one really needs to craft own code. Please use a YAML library such as [snakeyaml](https://bitbucket.org/snakeyaml/snakeyaml-engine/src/master).

If I understood correctly, you agree with creating a complete exporter only for CFF ? I can try to use a library and modify my exporter.

Reply:
> > `references` in CFF are mirrored in JabRef as `related`
> 
> Regarding this, you would like to parse the `references` field in both the importer and the exporter ?

Yes. Test cases are at https://github.com/citation-file-format/citation-file-format/blob/main/examples/1.2.0/pass/reference-conference-paper/CITATION.cff and other CFF files there.

Reply:
> If I understood correctly, you agree with creating a complete exporter only for CFF ? I can try to use a library and modify my exporter.

Yes. Please write down as [MADR](https://adr.github.io/madr/) document.

Reply:
> Then work on a proposal. We would like to see: the new syntax, an example, difficulties faced... Familiarize yourself with the [MADR format](https://adr.github.io/madr/) to systematically write down options and pros/cons.

I'll work on this, this project seems to be so much interesting. 
  
> Ah, now I see what you are aiming for.
> 
> In my mind, more than one CFF file would be created by the export. But this is not a typical use case. Typically, one exports multiple entries into one CFF.
> 
> Then, this whole relation thing is obsolete, because the exporter just uses software/dataset as topmost element and `references` for the others. The only exception is the `cites` relation. This can be used to determine `preferred-citation`. - Feels good. WDYT?

Yeah, feels good to me too. I'll try to implement the following: 
+ if one or more "non-software-nor-dataset" entries are exported, then I use the `references` field of CFF with a dummy topmost software element. 
+ if an entry contains a JabRef's `cites` attribute to another software entry, then I'll use this software entry as the topmost element and the CFF `preferred-citation` field for the main entry.
+ if an entry is a software or dataset element, then I'll simply use normal CFF conversion (e.g. `preferred-citation` if  and only if there is a JabRef's `cites` attributes)

> Please write down the exporter decisions in a new Markdown file in https://github.com/JabRef/jabref/tree/main/docs/code-howtos.

I'll do this. 

> Yes. - The preferred citation is stored as BibEntry having a cites on the "main" entry.

I feel like it would be useful that both entries have a cites on each other. This way, when one of them is exported to CFF, both of them are in the final file. WDYT ?

Reply:
> I feel like it would be useful that both entries have a cites on each other. This way, when one of them is exported to CFF, both of them are in the final file. WDYT ?

No. The semantics of `cites` is that an article cites another article (released earlier). If A cites B, then one can query A from the view of B by forEach'in all As and check if A cites B.

The relation is NOT [symmetric](https://en.wikipedia.org/wiki/Symmetric_relation). One should not force to make it one without reason!

Reply:
But this means that when a software will be imported then exported it will lose its `preferred-citation`.

Reply:
> The discussion was long. I did not see an update to my statement

I only had time to implement the change to use `snakeyaml` and support the `references` field. This is why this PR is still draft, I will implement your statements in the following days. 

> Thus, you say the complete opposite what I stated. Why? Why is it not possible to add the cites from the software?

This is possible, but I misunderstood what you stated above : 
> Yes. - The preferred citation is stored as BibEntry having a cites on the "main" entry. 

I was confused about this sentence, but now it is clear I'll add a `cites` from the software towards the `preferred-citation`. 

> I know when the user selects multiple entries, one struggles whether to use both the cites and the selected entries - or just the selected entries. - I would say: just the selected entries are exported.

If multiple "non-software-nor-dataset" entries are selected, then I'll use the `references` field with a dummy topmost software element. If several entries including a "software" one are selected, then I'll use this one as topmost element and the others as `references`, adding a `preferred-citation` for the `cites` element.  Finally, if several entries including several "software" ones are selected, then I'll use a dummy topmost element, and I'll do what you say here : just the selected entries are exported, as `references`. The `cites` won't be exported in this case.

Reply:
> For importing, there should be a "Merge entries" dialog popping up, where one can resolve issues in the cites field.

This will come later as I don't really know how to do that for the moment. Is this "Merge entries" window automatically popping up when several entries are created ?

Reply:
> This will come later as I don't really know how to do that for the moment. Is this "Merge entries" window automatically popping up when several entries are created ?

Just copy and paste an entry of the main table. Do the paste a second time: The popup will appear. -- No need to take care now, I think, JabRef handles. The only thing, JabRef will probably not handle is different content of the `cite` field and proper merge... Let's see.

Reply:
> This is why this PR is still draft, I will implement your statements in the following days.

Nice! If you want, you can update the PR description with check boxes what is still open.

Side comment: You did not create a user-documentation issue, did you?

Reply:
> Nice! If you want, you can update the PR description with check boxes what is still open.

I updated the PR description which check boxes still open.

Reply:
> Side comment: You did not create a user-documentation issue, did you?

I checked and there was this [issue](https://github.com/JabRef/user-documentation/issues/473), which is pretty the same but more general. It consisted in adding all Exporters name. The [page to complete](https://docs.jabref.org/collaborative-work/export), in user documentation, seems to ask for this issue to be solved.